File: Algorithms/p438_Find_All_Anagrams_in_a_String/p438_Find_All_Anagrams_in_a_String.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int n = s.size();
        int m = p.size();
        if(n<m) return {};
        vector<int> res;
        vector<int> cnt_p(256,0);
        for(int i=0;i<m;++i){
            ++cnt_p[p[i]];
        }
        vector<int> cnt(256,0);
        for(int i=0;i<m;++i){
            ++cnt[s[i]];
        }
        if(equal(cnt.begin(),cnt.end(),cnt_p.begin())){
            res.push_back(0);
        }
        for(int i=1;i<=n-m;++i){
            --cnt[s[i-1]];
            ++cnt[s[i+m-1]];
            if(equal(cnt.begin(),cnt.end(),cnt_p.begin())){
                res.push_back(i);
            }
        }
        return res;
    }
};

File: Algorithms/p438_Find_All_Anagrams_in_a_String/p438_Find_All_Anagrams_in_a_String.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        n = len(s)
        m = len(p)
        if n<m:
            return []
        res = []
        cnt_p = {chr(ord('a')+i):0 for i in range(26)}
        for c in p:
            cnt_p[c] += 1
        cnt = {chr(ord('a')+i):0 for i in range(26)}
        for i in range(0,m):
            cnt[s[i]] += 1
        if cnt==cnt_p:
            res.append(0)
        for i in range(1,n-m+1):
            cnt[s[i-1]] -= 1
            cnt[s[i+m-1]] += 1
            if cnt==cnt_p:
                res.append(i)
        return res
